From 57be37cf9edcdfa0586465f9d75c67e1c21c44e1 Mon Sep 17 00:00:00 2001
From: Konstantin Klubnichkin <kitsok@yandex-team.ru>
Date: Thu, 3 Nov 2022 16:14:25 +0000
Subject: [PATCH] Introduce yandex_ast2600 board

- Enable WDT3
- Enable PCIe EHCI controller
- Add DTS for yandex_ast2600 (AspeedTech dev board)
- Add DTS for yandex_som2600v1 (Yandex AST2600 SOM)
- Mask out reset of vital host-accessed peripherals
- Force NCSI disable in config file
- Enable raw and legacy SPL image support
- Add IBM config strings for FIT and signature support

%% original patch: 0001-Introduce-yandex_ast2600-board.patch
---
 arch/arm/dts/Makefile                        |   5 +-
 arch/arm/dts/yandex-evb2600.dts              | 316 ++++++++++++++++++
 arch/arm/dts/yandex-som2600v1.dts            | 319 +++++++++++++++++++
 arch/arm/mach-aspeed/ast2600/Kconfig         |   8 +
 arch/arm/mach-aspeed/ast2600/platform.S      |  58 ++++
 board/aspeed/yandex_ast2600/Kconfig          |  13 +
 board/aspeed/yandex_ast2600/Makefile         |   1 +
 board/aspeed/yandex_ast2600/yandex_ast2600.c | 200 ++++++++++++
 configs/yandex_ast2600_defconfig             | 153 +++++++++
 include/configs/yandex_ast2600.h             |  36 +++
 10 files changed, 1108 insertions(+), 1 deletion(-)
 create mode 100644 arch/arm/dts/yandex-evb2600.dts
 create mode 100644 arch/arm/dts/yandex-som2600v1.dts
 create mode 100644 board/aspeed/yandex_ast2600/Kconfig
 create mode 100644 board/aspeed/yandex_ast2600/Makefile
 create mode 100644 board/aspeed/yandex_ast2600/yandex_ast2600.c
 create mode 100644 configs/yandex_ast2600_defconfig
 create mode 100644 include/configs/yandex_ast2600.h

diff --git a/arch/arm/dts/Makefile b/arch/arm/dts/Makefile
index 6c34b83336..31af9dd697 100755
--- a/arch/arm/dts/Makefile
+++ b/arch/arm/dts/Makefile
@@ -695,7 +695,10 @@ dtb-$(CONFIG_ARCH_ASPEED) += \
 	ast2600-slt.dtb \
 	ast2600-tacoma.dtb \
 	ast2600-intel.dtb \
-	ast2600-tacoma.dtb
+	ast2600-tacoma.dtb \
+	yandex-evb2600.dtb \
+	yandex-som2600v1.dtb
+
 
 dtb-$(CONFIG_ARCH_STI) += stih410-b2260.dtb
 
diff --git a/arch/arm/dts/yandex-evb2600.dts b/arch/arm/dts/yandex-evb2600.dts
new file mode 100644
index 0000000000..188d9db62d
--- /dev/null
+++ b/arch/arm/dts/yandex-evb2600.dts
@@ -0,0 +1,316 @@
+/dts-v1/;
+
+#include "ast2600-u-boot.dtsi"
+
+/ {
+        model = "AST2600 EVB";
+        compatible = "aspeed,ast2600-evb", "aspeed,ast2600";
+
+	memory {
+		device_type = "memory";
+		reg = <0x80000000 0x40000000>;
+	};
+
+	chosen {
+		stdout-path = &uart5;
+	};
+
+	aliases {
+		mmc0 = &emmc_slot0;
+		mmc1 = &sdhci_slot0;
+		mmc2 = &sdhci_slot1;
+		spi0 = &fmc;
+		spi1 = &spi1;
+		spi2 = &spi2;
+		ethernet0 = &mac0;
+		ethernet1 = &mac1;
+		ethernet2 = &mac2;
+		ethernet3 = &mac3;
+	};
+
+	cpus {
+		cpu@0 {
+			clock-frequency = <800000000>;
+		};
+		cpu@1 {
+			clock-frequency = <800000000>;
+		};
+	};
+};
+
+&uart5 {
+	u-boot,dm-pre-reloc;
+	status = "okay";
+};
+
+&sdrammc {
+	clock-frequency = <400000000>;
+};
+
+&wdt1 {
+	status = "okay";
+};
+
+&wdt2 {
+	status = "okay";
+};
+
+&wdt3 {
+	status = "okay";
+};
+
+&mdio {
+	status = "okay";
+	pinctrl-names = "default";
+	pinctrl-0 = <	&pinctrl_mdio1_default &pinctrl_mdio2_default
+			&pinctrl_mdio3_default &pinctrl_mdio4_default>;
+	#address-cells = <1>;
+	#size-cells = <0>;
+	ethphy0: ethernet-phy@0 {
+		reg = <0>;
+	};
+
+	ethphy1: ethernet-phy@1 {
+		reg = <0>;
+	};
+
+	ethphy2: ethernet-phy@2 {
+		reg = <0>;
+	};
+
+	ethphy3: ethernet-phy@3 {
+		reg = <0>;
+	};
+};
+
+&mac0 {
+	status = "okay";
+	phy-mode = "rgmii-rxid";
+	phy-handle = <&ethphy0>;
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_rgmii1_default>;
+};
+
+&mac1 {
+	status = "okay";
+	phy-mode = "rgmii-rxid";
+	phy-handle = <&ethphy1>;
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_rgmii2_default>;
+};
+
+&mac2 {
+	status = "okay";
+	phy-mode = "rgmii";
+	phy-handle = <&ethphy2>;
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_rgmii3_default &pinctrl_mac3link_default>;
+};
+
+&mac3 {
+	status = "okay";
+	phy-mode = "rgmii";
+	phy-handle = <&ethphy3>;
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_rgmii4_default &pinctrl_mac4link_default>;
+};
+
+&fmc {
+	status = "okay";
+
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_fmcquad_default>;
+
+	flash@0 {
+		compatible = "spi-flash", "sst,w25q256";
+		status = "okay";
+		spi-max-frequency = <50000000>;
+		spi-tx-bus-width = <4>;
+		spi-rx-bus-width = <4>;
+	};
+
+	flash@1 {
+		compatible = "spi-flash", "sst,w25q256";
+		status = "okay";
+		spi-max-frequency = <50000000>;
+		spi-tx-bus-width = <4>;
+		spi-rx-bus-width = <4>;
+	};
+
+	flash@2 {
+		compatible = "spi-flash", "sst,w25q256";
+		status = "okay";
+		spi-max-frequency = <50000000>;
+		spi-tx-bus-width = <4>;
+		spi-rx-bus-width = <4>;
+	};
+};
+
+&spi1 {
+	status = "okay";
+
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_spi1_default &pinctrl_spi1abr_default
+			&pinctrl_spi1cs1_default &pinctrl_spi1wp_default
+			&pinctrl_spi1wp_default &pinctrl_spi1quad_default>;
+
+	flash@0 {
+		compatible = "spi-flash", "sst,w25q256";
+		status = "okay";
+		spi-max-frequency = <50000000>;
+		spi-tx-bus-width = <4>;
+		spi-rx-bus-width = <4>;
+	};
+
+	flash@1 {
+		compatible = "spi-flash", "sst,w25q256";
+		status = "okay";
+		spi-max-frequency = <50000000>;
+		spi-tx-bus-width = <4>;
+		spi-rx-bus-width = <4>;
+	};
+};
+
+&spi2 {
+	status = "okay";
+
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_spi2_default &pinctrl_spi2cs1_default
+			&pinctrl_spi2cs2_default &pinctrl_spi2quad_default>;
+
+	flash@0 {
+		compatible = "spi-flash", "sst,w25q256";
+		status = "okay";
+		spi-max-frequency = <50000000>;
+		spi-tx-bus-width = <4>;
+		spi-rx-bus-width = <4>;
+	};
+
+	flash@1 {
+		compatible = "spi-flash", "sst,w25q256";
+		status = "okay";
+		spi-max-frequency = <50000000>;
+		spi-tx-bus-width = <4>;
+		spi-rx-bus-width = <4>;
+	};
+
+	flash@2 {
+		compatible = "spi-flash", "sst,w25q256";
+		status = "okay";
+		spi-max-frequency = <50000000>;
+		spi-tx-bus-width = <4>;
+		spi-rx-bus-width = <4>;
+	};
+};
+
+&emmc {
+	u-boot,dm-pre-reloc;
+	timing-phase = <0x700ff>;
+};
+
+&emmc_slot0 {
+	u-boot,dm-pre-reloc;
+	status = "okay";
+	bus-width = <4>;
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_emmc_default>;
+	sdhci-drive-type = <1>;
+};
+
+&sdhci {
+	timing-phase = <0xc6ffff>;
+};
+
+&sdhci_slot0 {
+	status = "okay";
+	bus-width = <4>;
+	pwr-gpios = <&gpio0 ASPEED_GPIO(V, 0) GPIO_ACTIVE_HIGH>;
+	pwr-sw-gpios = <&gpio0 ASPEED_GPIO(V, 1) GPIO_ACTIVE_HIGH>;
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_sd1_default>;
+	sdhci-drive-type = <1>;
+};
+
+&sdhci_slot1 {
+	status = "okay";
+	bus-width = <4>;
+	pwr-gpios = <&gpio0 ASPEED_GPIO(V, 2) GPIO_ACTIVE_HIGH>;
+	pwr-sw-gpios = <&gpio0 ASPEED_GPIO(V, 3) GPIO_ACTIVE_HIGH>;
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_sd2_default>;
+	sdhci-drive-type = <1>;
+};
+
+&i2c4 {
+	status = "okay";
+
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_i2c5_default>;
+};
+
+&i2c5 {
+	status = "okay";
+
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_i2c6_default>;
+};
+
+&i2c6 {
+	status = "okay";
+
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_i2c7_default>;
+};
+
+&i2c7 {
+	status = "okay";
+
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_i2c8_default>;
+};
+
+&i2c8 {
+	status = "okay";
+
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_i2c9_default>;
+};
+
+#if 0
+&fsim0 {
+	status = "okay";
+};
+
+&fsim1 {
+	status = "okay";
+};
+#endif
+
+#if 0
+&ehci1 {
+	status = "okay";
+};
+#endif
+
+&display_port {
+	status = "okay";
+};
+
+&scu {
+	mac0-clk-delay = <0x10 0x0a
+			  0x10 0x10
+			  0x10 0x10>;
+	mac1-clk-delay = <0x10 0x0a
+			  0x10 0x10
+			  0x10 0x10>;
+	mac2-clk-delay = <0x08 0x04
+			  0x08 0x04
+			  0x08 0x04>;
+	mac3-clk-delay = <0x08 0x04
+			  0x08 0x04
+			  0x08 0x04>;
+};
+
+&hace {
+	status = "okay";
+};
diff --git a/arch/arm/dts/yandex-som2600v1.dts b/arch/arm/dts/yandex-som2600v1.dts
new file mode 100644
index 0000000000..587b8e1375
--- /dev/null
+++ b/arch/arm/dts/yandex-som2600v1.dts
@@ -0,0 +1,319 @@
+/dts-v1/;
+
+#include "ast2600-u-boot.dtsi"
+
+/ {
+        model = "AST2600 EVB";
+        compatible = "aspeed,ast2600-evb", "aspeed,ast2600";
+
+	memory {
+		device_type = "memory";
+		reg = <0x80000000 0x40000000>;
+	};
+
+	chosen {
+		stdout-path = &uart5;
+	};
+
+	aliases {
+		mmc0 = &emmc_slot0;
+		mmc1 = &sdhci_slot0;
+		mmc2 = &sdhci_slot1;
+		spi0 = &fmc;
+		spi1 = &spi1;
+		spi2 = &spi2;
+		ethernet0 = &mac2;
+		ethernet1 = &mac3;
+	};
+
+	cpus {
+		cpu@0 {
+			clock-frequency = <800000000>;
+		};
+		cpu@1 {
+			clock-frequency = <800000000>;
+		};
+	};
+};
+
+&uart5 {
+	u-boot,dm-pre-reloc;
+	status = "okay";
+};
+
+&sdrammc {
+	clock-frequency = <400000000>;
+};
+
+&wdt1 {
+	status = "okay";
+};
+
+&wdt2 {
+	status = "okay";
+};
+
+&wdt3 {
+	status = "okay";
+};
+
+&mdio {
+	status = "okay";
+	pinctrl-names = "default";
+	pinctrl-0 = <	&pinctrl_mdio1_default &pinctrl_mdio2_default
+			&pinctrl_mdio3_default &pinctrl_mdio4_default>;
+	#address-cells = <1>;
+	#size-cells = <0>;
+	ethphy0: ethernet-phy@0 {
+		reg = <0>;
+	};
+
+	ethphy1: ethernet-phy@1 {
+		reg = <0>;
+	};
+
+	ethphy2: ethernet-phy@2 {
+		reg = <0>;
+	};
+
+	ethphy3: ethernet-phy@3 {
+		reg = <0>;
+	};
+};
+
+&mac0 {
+	status = "disabled";
+	phy-mode = "rgmii-rxid";
+	phy-handle = <&ethphy0>;
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_rgmii1_default>;
+};
+
+&mac1 {
+	status = "disabled";
+	phy-mode = "rgmii-rxid";
+	phy-handle = <&ethphy1>;
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_rgmii2_default>;
+};
+
+&mac2 {
+	status = "okay";
+	phy-mode = "rgmii";
+	phy-handle = <&ethphy2>;
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_rgmii3_default &pinctrl_mac3link_default>;
+};
+
+&mac3 {
+	status = "okay";
+	phy-mode = "rgmii";
+	phy-handle = <&ethphy3>;
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_rgmii4_default &pinctrl_mac4link_default>;
+};
+
+&fmc {
+	status = "okay";
+
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_fmcquad_default>;
+
+	flash@0 {
+		compatible = "spi-flash", "sst,w25q256";
+		status = "okay";
+		spi-max-frequency = <50000000>;
+		spi-tx-bus-width = <4>;
+		spi-rx-bus-width = <4>;
+	};
+
+	flash@1 {
+		compatible = "spi-flash", "sst,w25q256";
+		status = "okay";
+		spi-max-frequency = <50000000>;
+		spi-tx-bus-width = <4>;
+		spi-rx-bus-width = <4>;
+	};
+
+	flash@2 {
+		compatible = "spi-flash", "sst,w25q256";
+		status = "okay";
+		spi-max-frequency = <50000000>;
+		spi-tx-bus-width = <4>;
+		spi-rx-bus-width = <4>;
+	};
+};
+
+&spi1 {
+	status = "okay";
+
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_spi1_default &pinctrl_spi1abr_default
+			&pinctrl_spi1cs1_default &pinctrl_spi1wp_default
+			&pinctrl_spi1wp_default &pinctrl_spi1quad_default>;
+
+	flash@0 {
+		compatible = "spi-flash", "sst,w25q256";
+		status = "okay";
+		spi-max-frequency = <50000000>;
+		spi-tx-bus-width = <4>;
+		spi-rx-bus-width = <4>;
+	};
+
+	flash@1 {
+		compatible = "spi-flash", "sst,w25q256";
+		status = "okay";
+		spi-max-frequency = <50000000>;
+		spi-tx-bus-width = <4>;
+		spi-rx-bus-width = <4>;
+	};
+};
+
+&spi2 {
+	status = "okay";
+
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_spi2_default &pinctrl_spi2cs1_default
+			&pinctrl_spi2cs2_default &pinctrl_spi2quad_default>;
+
+	flash@0 {
+		compatible = "spi-flash", "sst,w25q256";
+		status = "okay";
+		spi-max-frequency = <50000000>;
+		spi-tx-bus-width = <4>;
+		spi-rx-bus-width = <4>;
+	};
+
+	flash@1 {
+		compatible = "spi-flash", "sst,w25q256";
+		status = "okay";
+		spi-max-frequency = <50000000>;
+		spi-tx-bus-width = <4>;
+		spi-rx-bus-width = <4>;
+	};
+
+	flash@2 {
+		compatible = "spi-flash", "sst,w25q256";
+		status = "okay";
+		spi-max-frequency = <50000000>;
+		spi-tx-bus-width = <4>;
+		spi-rx-bus-width = <4>;
+	};
+};
+
+&emmc {
+	u-boot,dm-pre-reloc;
+	timing-phase = <0x700ff>;
+};
+
+&emmc_slot0 {
+	u-boot,dm-pre-reloc;
+	status = "okay";
+	bus-width = <4>;
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_emmc_default>;
+	sdhci-drive-type = <1>;
+};
+
+&sdhci {
+	timing-phase = <0xc6ffff>;
+};
+
+&sdhci_slot0 {
+	status = "okay";
+	bus-width = <4>;
+	pwr-gpios = <&gpio0 ASPEED_GPIO(V, 0) GPIO_ACTIVE_HIGH>;
+	pwr-sw-gpios = <&gpio0 ASPEED_GPIO(V, 1) GPIO_ACTIVE_HIGH>;
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_sd1_default>;
+	sdhci-drive-type = <1>;
+};
+
+&sdhci_slot1 {
+	status = "okay";
+	bus-width = <4>;
+	pwr-gpios = <&gpio0 ASPEED_GPIO(V, 2) GPIO_ACTIVE_HIGH>;
+	pwr-sw-gpios = <&gpio0 ASPEED_GPIO(V, 3) GPIO_ACTIVE_HIGH>;
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_sd2_default>;
+	sdhci-drive-type = <1>;
+};
+
+&i2c4 {
+	status = "okay";
+
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_i2c5_default>;
+};
+
+&i2c5 {
+	status = "okay";
+
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_i2c6_default>;
+};
+
+&i2c6 {
+	status = "okay";
+
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_i2c7_default>;
+};
+
+&i2c7 {
+	status = "okay";
+
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_i2c8_default>;
+};
+
+&i2c8 {
+	status = "okay";
+
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_i2c9_default>;
+};
+
+#if 0
+&fsim0 {
+	status = "okay";
+};
+
+&fsim1 {
+	status = "okay";
+};
+#endif
+
+#if 0
+&ehci1 {
+	status = "okay";
+};
+#endif
+
+&display_port {
+	status = "okay";
+};
+
+&scu {
+	mac0-clk-delay = <0x10 0x0a
+			  0x10 0x10
+			  0x10 0x10>;
+	mac1-clk-delay = <0x10 0x0a
+			  0x10 0x10
+			  0x10 0x10>;
+	mac2-clk-delay = <0x08 0x04
+			  0x08 0x04
+			  0x08 0x04>;
+	mac3-clk-delay = <0x08 0x04
+			  0x08 0x04
+			  0x08 0x04>;
+};
+
+&hace {
+	u-boot,dm-pre-reloc;
+	status = "okay";
+};
+
+&gpio0 {
+	u-boot,dm-pre-reloc;
+};
diff --git a/arch/arm/mach-aspeed/ast2600/Kconfig b/arch/arm/mach-aspeed/ast2600/Kconfig
index f5852afa77..3e7d5a2c6b 100644
--- a/arch/arm/mach-aspeed/ast2600/Kconfig
+++ b/arch/arm/mach-aspeed/ast2600/Kconfig
@@ -51,6 +51,13 @@ config TARGET_QUALCOMM_DC_SCM_V1
 	  QUALCOMM-DC-SCM-V1 is a Qualcomm DC-SCM V1 board which is
 	  equipped with AST2600.
 
+config TARGET_YANDEX_AST2600
+	bool "YANDEX-AST2600"
+	depends on ASPEED_AST2600
+	help
+	  YANDEX_AST2600 is an Yandex mods to U-Boot.
+	  AST2600A3 EVB as test hardware.
+
 endchoice
 
 source "board/aspeed/evb_ast2600/Kconfig"
@@ -59,5 +66,6 @@ source "board/aspeed/slt_ast2600/Kconfig"
 source "board/aspeed/ast2600_ibm/Kconfig"
 source "board/aspeed/ast2600_intel/Kconfig"
 source "board/qualcomm/dc-scm-v1/Kconfig"
+source "board/aspeed/yandex_ast2600/Kconfig"
 
 endif
diff --git a/arch/arm/mach-aspeed/ast2600/platform.S b/arch/arm/mach-aspeed/ast2600/platform.S
index 6623d58957..16531f4098 100644
--- a/arch/arm/mach-aspeed/ast2600/platform.S
+++ b/arch/arm/mach-aspeed/ast2600/platform.S
@@ -59,6 +59,19 @@
 #define AST_FMC_WDT1_CTRL_MODE	(AST_FMC_BASE + 0x060)
 #define AST_FMC_WDT2_CTRL_MODE	(AST_FMC_BASE + 0x064)
 
+#define AST_WDT1_BASE			0x1E785000
+#define AST_WDT1_RESET_MASK1	(AST_WDT1_BASE + 0x01C)
+#define AST_WDT1_RESET_MASK2	(AST_WDT1_BASE + 0x020)
+#define AST_WDT2_BASE			0x1E785040
+#define AST_WDT2_RESET_MASK1	(AST_WDT2_BASE + 0x01C)
+#define AST_WDT2_RESET_MASK2	(AST_WDT2_BASE + 0x020)
+#define AST_WDT3_BASE			0x1E785080
+#define AST_WDT3_RELOAD_VALUE	(AST_WDT3_BASE + 0x04)
+#define AST_WDT3_CNT_RESTART	(AST_WDT3_BASE + 0x08)
+#define AST_WDT3_CTRL			(AST_WDT3_BASE + 0x0C)
+#define AST_WDT3_RESET_MASK1	(AST_WDT3_BASE + 0x01C)
+#define AST_WDT3_RESET_MASK2	(AST_WDT3_BASE + 0x020)
+
 #define AST_GPIO_BASE		(0x1E780000)
 #define AST_GPIOYZ_DATA_VALUE	(AST_GPIO_BASE + 0x1E0)
 
@@ -247,6 +260,51 @@ wait_lock:
 	ldr	r1, =AST_SEC_VAULT_KEY_CTRL
 	str	r0, [r1]
 
+	/* disable eSPI, LPC and PWM resets on WDT1 reset */
+	ldr r0, =AST_WDT1_RESET_MASK1
+	ldr r1, [r0]
+	ldr r2, =0x00000804 // Mask out reset of AHB bridges, Video engine
+	bic r1, r2
+	str r1, [r0]
+	ldr r0, =AST_WDT1_RESET_MASK2
+	ldr r1, [r0]
+	ldr r2, =0x04002804 // Mask out reset of eSPI, PWM, LPC, AHB2
+	bic r1, r2
+	str r1, [r0]
+
+	/* disable eSPI, LPC and PWM resets on WDT2 reset */
+	ldr r0, =AST_WDT2_RESET_MASK1
+	ldr r1, [r0]
+	ldr r2, =0x00000804 // Mask out reset of AHB bridges, Video engine
+	bic r1, r2
+	str r1, [r0]
+	ldr r0, =AST_WDT2_RESET_MASK2
+	ldr r1, [r0]
+	ldr r2, =0x04002804
+	bic r1, r2
+	str r1, [r0]
+
+	/* Set up WDT3 to 10 seconds */
+	ldr r0, =AST_WDT3_RESET_MASK1
+	ldr r1, [r0]
+	ldr r2, =0x00000804 // Mask out reset of AHB bridges, Video engine
+	bic r1, r2
+	str r1, [r0]
+	ldr r0, =AST_WDT3_RESET_MASK2
+	ldr r1, [r0]
+	ldr r2, =0x04002804
+	bic r1, r2
+	str r1, [r0]
+	ldr r0, =AST_WDT3_RELOAD_VALUE
+	ldr r1, =(1000000 * 10)
+	str r1, [r0]
+	ldr r0, =AST_WDT3_CNT_RESTART
+	ldr r1, =0x4755
+	str r1, [r0]
+	ldr r0, =AST_WDT3_CTRL
+	ldr r1, =0x13
+	str r1, [r0]
+
 	/* PCIeRC/E2M8 power-on reset comes from SCU040
 	It need set SCU040[18] high to reset PCIeRC/E2M
 	when AC power-on */
diff --git a/board/aspeed/yandex_ast2600/Kconfig b/board/aspeed/yandex_ast2600/Kconfig
new file mode 100644
index 0000000000..276af705ec
--- /dev/null
+++ b/board/aspeed/yandex_ast2600/Kconfig
@@ -0,0 +1,13 @@
+if TARGET_YANDEX_AST2600
+
+config SYS_BOARD
+	default "yandex_ast2600"
+
+config SYS_VENDOR
+	default "aspeed"
+
+config SYS_CONFIG_NAME
+	string "board configuration name"
+	default "yandex_ast2600"
+
+endif
diff --git a/board/aspeed/yandex_ast2600/Makefile b/board/aspeed/yandex_ast2600/Makefile
new file mode 100644
index 0000000000..1c668fe88f
--- /dev/null
+++ b/board/aspeed/yandex_ast2600/Makefile
@@ -0,0 +1 @@
+obj-y += yandex_ast2600.o
diff --git a/board/aspeed/yandex_ast2600/yandex_ast2600.c b/board/aspeed/yandex_ast2600/yandex_ast2600.c
new file mode 100644
index 0000000000..f9d84f6258
--- /dev/null
+++ b/board/aspeed/yandex_ast2600/yandex_ast2600.c
@@ -0,0 +1,200 @@
+// SPDX-License-Identifier: GPL-2.0+
+/*
+ * Copyright (C) ASPEED Technology Inc.
+ */
+#include <common.h>
+#include <asm/io.h>
+#include <dm.h>
+#include <dm/uclass.h>
+#include <i2c.h>
+#include <watchdog.h>
+
+DECLARE_GLOBAL_DATA_PTR;
+
+#define AST_GPIO_BASE		(0x1E780000)
+#define AST_GPIOABCD_DRCTN	(AST_GPIO_BASE + 0x004)
+#define AST_GPIOEFGH_DRCTN	(AST_GPIO_BASE + 0x024)
+#define AST_GPIOMNOP_DRCTN	(AST_GPIO_BASE + 0x07C)
+#define AST_GPIOQRST_DRCTN	(AST_GPIO_BASE + 0x084)
+#define AST_GPIOUVWX_DRCTN	(AST_GPIO_BASE + 0x08C)
+#define AST_GPIOYZ_DRCTN	(AST_GPIO_BASE + 0x1E4)
+
+int board_init(void)
+{
+	u32 direction;
+
+	struct udevice *dev;
+	int i;
+	int ret;
+
+	gd->bd->bi_boot_params = CONFIG_SYS_SDRAM_BASE + 0x100;
+
+	/*
+	 * Loop over all MISC uclass drivers to call the comphy code
+	 * and init all CP110 devices enabled in the DT
+	 */
+	i = 0;
+	while (1) {
+		/* Call the comphy code via the MISC uclass driver */
+		ret = uclass_get_device(UCLASS_MISC, i++, &dev);
+
+		/* We're done, once no further CP110 device is found */
+		if (ret)
+			break;
+	}
+	//return 0;
+	/*
+	 * in FT/SLT board, 4 MDIO channels are all connected to a single PHY
+	 * chip.  So only one MDIO channel can access the PHY chip at a time and
+	 * the pins of the channels that are not under testing will be set to
+	 * GPIO input.  So simply the test program in Linux kernel, we configure
+	 * GPIO setting here.
+	 *
+	 * GPIOS[1:0] -> MDC1 & MDIO1
+	 * GPIOB[5:4] -> MDC2 & MDIO2
+	 * GPIOA[1:0] -> MDC3 & MDIO3
+	 * GPIOA[3:2] -> MDC4 & MDIO4
+	*/
+	/* GPIOS[1:0] */
+	direction = readl(AST_GPIOQRST_DRCTN);
+	direction &= ~GENMASK(17, 16);
+	writel(direction, AST_GPIOQRST_DRCTN);
+
+	/* GPIOA[3:0] and GPIOB[5:4] */
+	direction = readl(AST_GPIOABCD_DRCTN);
+	direction &= ~(GENMASK(3, 0) | GENMASK(13, 12));
+	writel(direction, AST_GPIOABCD_DRCTN);
+
+	/*
+	 * set 32 GPIO ouput pins for ATE report
+	 *   GPIOV[7:0] -> ATE[7:0]
+	 *   GPIOY[3:0] -> ATE[11:8]
+	 *   GPIOM[3:0] -> ATE[15:12]
+	 *   GPIOH[3:0] -> ATE[19:16]
+	 *   GPIOB[3:0] -> ATE[23:20]
+	 *   GPIOM[5:4] -> ATE[25:24]
+	 *   GPION[5:0] -> ATE[31:26]
+	 */
+	/* GPIOB[3:0] */
+	direction = readl(AST_GPIOABCD_DRCTN);
+	direction |= 0xF00;
+	writel(direction, AST_GPIOABCD_DRCTN);
+
+	/* GPIOH[3:0] */
+	direction = readl(AST_GPIOEFGH_DRCTN);
+	direction |= 0xF000000;
+	writel(direction, AST_GPIOEFGH_DRCTN);
+
+	/* GPIOM[3:0], GPIOM[5:4], GPION[5:0] */
+	direction = readl(AST_GPIOMNOP_DRCTN);
+	direction |= 0x3F3F;
+	writel(direction, AST_GPIOMNOP_DRCTN);
+
+	/* GPIOV[7:0] */
+	direction = readl(AST_GPIOUVWX_DRCTN);
+	direction |= 0xFF00;
+	writel(direction, AST_GPIOUVWX_DRCTN);
+
+	/* GPIOY[3:0] */
+	direction = readl(AST_GPIOYZ_DRCTN);
+	direction |= 0xF;
+	writel(direction, AST_GPIOYZ_DRCTN);
+
+	return 0;
+}
+
+/*
+ * Copyright 2021 Yandex LLC
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version
+ * 2 of the License, or (at your option) any later version.
+ */
+
+
+#define AST_WDT1_BASE 0x1E785000
+#define AST_WDT2_BASE 0x1E785040
+#define AST_WDT3_BASE 0x1E785080
+#define AST_WDT4_BASE 0x1E7850C0
+#define AST_WDT5_BASE 0x1E785100
+#define AST_WDT6_BASE 0x1E785140
+#define AST_WDT7_BASE 0x1E785180
+#define AST_WDT8_BASE 0x1E7851C0
+#define AST_SCU510_CLR 0x1E6E2514
+
+#define WDT3_TIMEOUT 180
+
+#define SCU040	0x1e6e2040
+#define SCU044	0x1e6e2044
+#define SCU080	0x1e6e2080
+#define SCU440	0x1e6e2440
+#define SCUC20	0x1e6e2c20
+#define HUB00	0x1e6ea000
+
+/* Init WDT3, set timeout */
+void init_wdt3(u32 seconds)
+{
+	printf("Disable all WDTs\n");
+	writel((u32) 0x00, AST_WDT1_BASE + 0x0c);
+	writel((u32) 0x00, AST_WDT2_BASE + 0x0c);
+	writel((u32) 0x00, AST_WDT4_BASE + 0x0c);
+	writel((u32) 0x00, AST_WDT5_BASE + 0x0c);
+	writel((u32) 0x00, AST_WDT6_BASE + 0x0c);
+	writel((u32) 0x00, AST_WDT7_BASE + 0x0c);
+	writel((u32) 0x00, AST_WDT8_BASE + 0x0c);
+	printf("Disable ABR\n");
+	writel((u32) 0x400800, AST_SCU510_CLR);
+
+	printf("WDT3 enabled, %i seconds\n", seconds);
+	writel((u32) 1000000 * seconds, AST_WDT3_BASE + 0x04);
+	writel(0x4755, AST_WDT3_BASE + 0x08);
+	writel(0x13, AST_WDT3_BASE + 0x0c);
+}
+
+/* Enable PCIe EHCI as descriped
+ * on p.106 of SDK user guide */
+void init_pcie_ehci(void)
+{
+	u32 reg;
+	reg = readl(SCU440);
+	reg &= ~(0x03 << 24); reg = 0x20000000;
+	writel(reg, SCU440);
+
+	reg = readl(SCU040);
+	reg |= (1<<14); reg = 0x4000;
+	writel(reg, SCU040);
+
+	reg = readl(SCU080);
+	reg |= (1<<14); reg = 0x4000;
+	writel(reg, SCU080);
+
+	/* Here should be 10ms delay */
+	printf("Initializing PCIe EHCI\n");
+
+	reg = readl(SCU044);
+	reg |= (1<<14); reg = 0x4000;
+	writel(reg, SCU044);
+
+	reg = readl(SCUC20);
+	reg |= (1<<16); reg = 0xd007b;
+	writel(reg, SCUC20);
+
+	reg = readl(HUB00); // 18.4.1
+	reg |= (1<<11); reg = 0x80070801;
+	writel(reg, HUB00);
+}
+
+int board_late_init(void)
+{
+	printf("Yandex late init routine\n");
+	//init_pcie_ehci();
+	init_wdt3(WDT3_TIMEOUT);
+	return 0;
+}
+
+int board_early_init_f(void)
+{
+	return 0;
+}
+
diff --git a/configs/yandex_ast2600_defconfig b/configs/yandex_ast2600_defconfig
new file mode 100644
index 0000000000..2fc5144cc4
--- /dev/null
+++ b/configs/yandex_ast2600_defconfig
@@ -0,0 +1,153 @@
+CONFIG_ARM=y
+CONFIG_SYS_CONFIG_NAME="yandex_ast2600"
+CONFIG_SYS_DCACHE_OFF=y
+CONFIG_POSITION_INDEPENDENT=y
+CONFIG_SPL_SYS_THUMB_BUILD=y
+CONFIG_SYS_THUMB_BUILD=y
+CONFIG_SPL_LDSCRIPT="arch/$(ARCH)/mach-aspeed/ast2600/u-boot-spl.lds"
+CONFIG_ARCH_ASPEED=y
+CONFIG_SYS_TEXT_BASE=0x10000
+CONFIG_ASPEED_AST2600=y
+CONFIG_TARGET_YANDEX_AST2600=y
+CONFIG_SPL_LIBCOMMON_SUPPORT=y
+CONFIG_SPL_LIBGENERIC_SUPPORT=y
+CONFIG_SYS_MALLOC_F_LEN=0x2000
+CONFIG_SPL_MMC_SUPPORT=y
+CONFIG_SPL_SERIAL_SUPPORT=y
+CONFIG_SPL_DRIVERS_MISC_SUPPORT=y
+CONFIG_ENV_SIZE=0x10000
+CONFIG_ENV_OFFSET=0xE0000
+CONFIG_SPL=y
+CONFIG_SPL_STACK_R_ADDR=0x93000000
+CONFIG_ARMV7_BOOT_SEC_DEFAULT=y
+CONFIG_ARMV7_PSCI_NR_CPUS=2
+CONFIG_NR_DRAM_BANKS=1
+CONFIG_FIT=y
+CONFIG_USE_BOOTARGS=y
+CONFIG_BOOTARGS="console=ttyS4,115200n8 root=/dev/ram rw"
+CONFIG_USE_BOOTCOMMAND=y
+CONFIG_BOOTCOMMAND="bootm 20100000"
+CONFIG_SYS_CONSOLE_ENV_OVERWRITE=y
+CONFIG_DISPLAY_BOARDINFO_LATE=y
+CONFIG_ARCH_EARLY_INIT_R=y
+CONFIG_BOARD_EARLY_INIT_F=y
+CONFIG_SPL_RAW_IMAGE_SUPPORT=y
+CONFIG_SPL_LEGACY_IMAGE_SUPPORT=y
+CONFIG_SPL_SYS_MALLOC_SIMPLE=y
+CONFIG_SPL_STACK_R=y
+CONFIG_SPL_SEPARATE_BSS=y
+CONFIG_SPL_DM_RESET=y
+CONFIG_SPL_RAM_SUPPORT=y
+CONFIG_SPL_RAM_DEVICE=y
+CONFIG_SPL_WATCHDOG_SUPPORT=y
+CONFIG_SPL_YMODEM_SUPPORT=y
+CONFIG_HUSH_PARSER=y
+# CONFIG_AUTO_COMPLETE is not set
+CONFIG_SYS_PROMPT="ast# "
+CONFIG_CMD_BOOTZ=y
+# CONFIG_CMD_ELF is not set
+# CONFIG_CMD_IMI is not set
+# CONFIG_CMD_XIMG is not set
+CONFIG_CMD_MEMTEST=y
+CONFIG_SYS_ALT_MEMTEST=y
+CONFIG_CMD_CLK=y
+CONFIG_CMD_GPIO=y
+CONFIG_CMD_GPT=y
+# CONFIG_RANDOM_UUID is not set
+CONFIG_CMD_I2C=y
+CONFIG_CMD_MMC=y
+CONFIG_CMD_PART=y
+CONFIG_CMD_SF=y
+CONFIG_CMD_DHCP=y
+CONFIG_CMD_MII=y
+CONFIG_CMD_PING=y
+CONFIG_CMD_EXT2=y
+CONFIG_CMD_EXT4=y
+CONFIG_CMD_EXT4_WRITE=y
+CONFIG_CMD_FAT=y
+CONFIG_CMD_FS_GENERIC=y
+CONFIG_CMD_MTDPARTS=y
+# CONFIG_SPL_EFI_PARTITION is not set
+CONFIG_SPL_OF_CONTROL=y
+CONFIG_ENV_IS_IN_SPI_FLASH=y
+CONFIG_USE_ENV_SPI_BUS=y
+CONFIG_ENV_SPI_BUS=0
+CONFIG_USE_ENV_SPI_CS=y
+CONFIG_ENV_SPI_CS=0
+CONFIG_USE_ENV_SPI_MAX_HZ=y
+CONFIG_ENV_SPI_MAX_HZ=100000000
+CONFIG_NET_RANDOM_ETHADDR=y
+CONFIG_SPL_DM=y
+CONFIG_REGMAP=y
+CONFIG_SYSCON=y
+CONFIG_SPL_OF_TRANSLATE=y
+CONFIG_CLK=y
+CONFIG_SPL_CLK=y
+CONFIG_DM_GPIO=y
+CONFIG_ASPEED_GPIO=y
+CONFIG_DM_I2C=y
+CONFIG_SYS_I2C_ASPEED=y
+CONFIG_MISC=y
+CONFIG_ASPEED_AHBC=y
+CONFIG_DM_MMC=y
+CONFIG_SPL_MMC_TINY=y
+CONFIG_MMC_SDHCI=y
+CONFIG_MMC_SDHCI_ASPEED=y
+CONFIG_DM_SPI_FLASH=y
+CONFIG_SPI_FLASH=y
+CONFIG_SPI_FLASH_GIGADEVICE=y
+CONFIG_SPI_FLASH_MACRONIX=y
+CONFIG_SPI_FLASH_SPANSION=y
+CONFIG_SPI_FLASH_STMICRO=y
+CONFIG_SPI_FLASH_WINBOND=y
+CONFIG_PHY_BROADCOM=y
+CONFIG_PHY_REALTEK=y
+CONFIG_DM_ETH=y
+CONFIG_PHY_GIGE=y
+CONFIG_FTGMAC100=y
+CONFIG_MDIO=y
+CONFIG_PHY=y
+CONFIG_PINCTRL=y
+CONFIG_RAM=y
+CONFIG_SPL_RAM=y
+CONFIG_DM_SERIAL=y
+CONFIG_SYS_NS16550=y
+CONFIG_SPI=y
+CONFIG_DM_SPI=y
+CONFIG_SYSRESET=y
+CONFIG_WDT=y
+CONFIG_USE_TINY_PRINTF=y
+CONFIG_TPM=y
+CONFIG_SPL_TPM=y
+# CONFIG_EFI_LOADER is not set
+CONFIG_BOARD_LATE_INIT=y
+CONFIG_ASPEED_BYPASS_SELFTEST=y
+CONFIG_ASPEED_UDC=y
+CONFIG_USB=y
+CONFIG_USB_GADGET=y
+CONFIG_USB_DEVICE=y
+CONFIG_DM_USB=y
+CONFIG_ASPEED_DP=y
+CONFIG_PHY_NCSI=n
+
+# IBM P10BMC
+CONFIG_ASPEED_HACE_V1=y
+CONFIG_ENV_OFFSET=0xE0000
+CONFIG_FIT_ENABLE_SHA384_SUPPORT=y
+CONFIG_FIT_ENABLE_SHA512_SUPPORT=y
+CONFIG_FIT_SIGNATURE=y
+CONFIG_SPL_BOARD_INIT=y
+CONFIG_SPL_FIT_IMAGE_TINY=y
+CONFIG_SPL_FIT_SIGNATURE=y
+CONFIG_SPL_GPIO_HOG=y
+CONFIG_SPL_GPIO_SUPPORT=y
+CONFIG_SPL_LOAD_FIT=y
+CONFIG_SPL_SHA1_SUPPORT=y
+CONFIG_SPL_SHA256_SUPPORT=y
+CONFIG_SPL_SHA384_SUPPORT=y
+CONFIG_SPL_SHA512_SUPPORT=y
+CONFIG_SPL_STACK_R_ADDR=0x90300000
+CONFIG_SPL_SYS_MALLOC_F_LEN=0x1000
+CONFIG_SPL_TINY_MEMSET=y
+
+CONFIG_ASPEED_ECC=y
diff --git a/include/configs/yandex_ast2600.h b/include/configs/yandex_ast2600.h
new file mode 100644
index 0000000000..f4f1b46d89
--- /dev/null
+++ b/include/configs/yandex_ast2600.h
@@ -0,0 +1,36 @@
+/* SPDX-License-Identifier: GPL-2.0+ */
+/*
+ * Copyright (C) ASPEED Technology Inc.
+ *
+ */
+
+#ifndef __CONFIG_H
+#define __CONFIG_H
+
+#include <configs/aspeed-common.h>
+
+#define CONFIG_SYS_MEMTEST_START	(CONFIG_SYS_SDRAM_BASE + 0x300000)
+#define CONFIG_SYS_MEMTEST_END		(CONFIG_SYS_MEMTEST_START + 0x5000000)
+
+#define CONFIG_SYS_UBOOT_BASE		CONFIG_SYS_TEXT_BASE
+
+/* Memory Info */
+#define CONFIG_SYS_LOAD_ADDR		0x83000000
+
+#ifdef CONFIG_SPL_TINY
+#ifdef CONFIG_SPL_BUILD
+#define CONFIG_SYS_NS16550_REG_SIZE 2
+#endif
+#endif
+
+/* SPL */
+#define CONFIG_SPL_TEXT_BASE		0x00000000
+#define CONFIG_SPL_MAX_SIZE		0x00010000
+#define CONFIG_SPL_STACK		0x10016000
+#define CONFIG_SPL_BSS_START_ADDR	0x90000000
+#define CONFIG_SPL_BSS_MAX_SIZE		0x00100000
+#define CONFIG_SPL_LOAD_FIT_ADDRESS     0x00010000
+
+#define CONFIG_SUPPORT_EMMC_BOOT
+
+#endif	/* __CONFIG_H */
-- 
2.36.0

